triedb/pathdb/states.go

```diff
@@ -43,9 +49,528 @@ func (c *counter) report(count metrics.Meter, size metrics.Meter) {
+// for the given account. If the whole storage is destructed in this layer, then
+// an additional flag *destructed = true* will be returned, otherwise the flag is
+// false. Besides, the returned list will include the hash of deleted storage slot.
+// Note a special case is an account is deleted in a prior tx but is recreated in
+// the following tx with some storage slots set. In this case the returned list is
+// not empty but the flag is true.
+//
+// Note, the returned slice is not a copy, so do not modify it.
+//
+//nolint:unused
+func (s *stateSet) storageList(accountHash common.Hash) ([]common.Hash, bool) {
+	s.lock.RLock()
+	_, destructed := s.destructSet[accountHash]
+	if _, ok := s.storageData[accountHash]; !ok {
+		// Account not tracked by this layer
+		s.lock.RUnlock()
+		return nil, destructed
+	}
+	// If an old list already exists, return it
+	if list, exist := s.storageListSorted[accountHash]; exist {
+		s.lock.RUnlock()
+		return list, destructed // the cached list can't be nil
+	}
+	s.lock.RUnlock()
+
+	// No old sorted account list exists, generate a new one
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	storageList := maps.Keys(s.storageData[accountHash])
+	slices.SortFunc(storageList, common.Hash.Cmp)
+	s.storageListSorted[accountHash] = storageList
+	return storageList, destructed
+}
+
+// clearCache invalidates the cached account list and storage lists.
+func (s *stateSet) clearCache() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.accountListSorted = nil
+	s.storageListSorted = make(map[common.Hash][]common.Hash)
+}
+
+// merge integrates the accounts and storages from the external set into the
+// local set, ensuring the combined set reflects the combined state of both.
+//
+// The provided state set will remain unchanged, as it may still be referenced
+// by other layers.
+func (s *stateSet) merge(set *stateSet) {
+	var (
+		delta             int
+		accountOverwrites counter
+		storageOverwrites counter
+		destructs         []destruct
+	)
+	// Apply account deletion markers and discard any previously cached data if exists
+	for accountHash := range set.destructSet {
+		if origin, ok := s.accountData[accountHash]; ok {
+			delta -= common.HashLength + len(origin)
+			accountOverwrites.add(common.HashLength + len(origin))
+			delete(s.accountData, accountHash)
+		}
+		if _, ok := s.storageData[accountHash]; ok {
+			// Looping through the nested map may cause slight performance degradation.
+			// However, since account destruction is no longer possible after the cancun
+			// fork, this overhead is considered acceptable.
+			for _, val := range s.storageData[accountHash] {
+				delta -= 2*common.HashLength + len(val)
+				storageOverwrites.add(2*common.HashLength + len(val))
+			}
+			delete(s.storageData, accountHash)
+		}
+		// Keep track of whether the account has already been marked as destructed.
+		// This additional marker is useful for undoing the merge operation.
+		_, exist := s.destructSet[accountHash]
```

Comment:
How can it already be marked as destructed? I guess the only case would be
1. `tx 1` creates contract at address `C`  via `CREATE2`
2. `tx 2` self-destructs `C`
3. `tx 3` creates contract at address `C`  via `CREATE2`
4. `tx 4` self-destructs `C`
5. ... and so on ...

Is this more or less the scenario that make you do the whole journalling thingy? Or are there other cornercases?

Reply:
These two sets belong to different blocks.

In block A, the account x was destructed, and then resurrected;
In block B, the account x was destructed again;

It's a valid operation before the Cancun fork (in which the self-destruction is disabled).

Reply:
the destruct set has been removed